src/BTrees/_datatypes.py

```diff
@@ -359,22 +385,129 @@ class Q(_AbstractUIntDataType):
     """
     __slots__ = ()
-    prefix_code = 'fs'
+    tree_size = 500
     default_bucket_size = 500
 
     def __init__(self, length):
-        super(Bytes, self).__init__()
+        super(_AbstractBytes, self).__init__()
         self._length = length
 
     def __call__(self, item):
         if not isinstance(item, bytes) or len(item) != self._length:
-            raise TypeError("%s-byte array expected" % self._length)
+            raise TypeError("%s-byte array expected, not %r" % (self._length, item))
         return item
 
     def supports_value_union(self):
         return False
+
+
+class f(_AbstractBytes):
+    """
+    The key type for an ``fs`` tree.
+
+    This is a two-byte prefix of an overall 8-byte value
+    like a ZODB object ID or transaction ID.
+    """
+
+    # Our keys are treated like integers; the module
+    # implements IIntegerObjectBTreeModule
+    long_name = 'Integer'
+    prefix_code = 'f'
+
+    def __init__(self):
+        _AbstractBytes.__init__(self, 2)
+
+    def supports_value_union(self):
+        # We don't implement 'multiunion' for fsBTree.
+        return False
+
+    # Check it can be converted to a two-byte
+    # value. Note that even though we allow negative values
+    # that can break test assumptions: -1 < 0 < 1, but the byte
+    # values for those are \xff\xff > \x00\x00 < \x00\x01.
+    _as_2_bytes = Struct('>h').pack
+
+    def coerce(self, item):
+        try:
+            return self(item)
+        except TypeError:
+            try:
+                return self._as_2_bytes(object_to_index(item))
+            except struct_error as e:
+                raise TypeError(e)
+
+    def make_toString(self):
+        def toString(self):
+            return b''.join(self._keys) + b''.join(self._values)
+        return toString
+
+    def make_fromString(self):
+        def fromString(self, v):
+            length = len(v)
+            if length % 8 != 0:
+                raise ValueError()
+            count = length // 8
+            keys, values = v[:count*2], v[count*2:]
+            self.clear()
+            while keys and values:
+                key, keys = keys[:2], keys[2:]
+                value, values = values[:6], values[6:]
+                self._keys.append(key)
+                self._values.append(value)
+            return self
+        return fromString
+
+    def add_extra_methods(self, base_name, cls):
+        if base_name == 'Bucket':
+            cls.toString = self.make_toString()
+            cls.fromString = self.make_fromString()
+
+class s(_AbstractBytes):
+    """
+    The value type for an ``fs`` tree.
+
+    This is a 6-byte suffix of an overall 8-byte value
+    like a ZODB object ID or transaction ID.
+    """
+
+    # Our values are treated like objects; the
+    # module implements IIntegerObjectBTreeModule
+    long_name = 'Object'
+    prefix_code = 's'
+
+
+    def __init__(self):
+        _AbstractBytes.__init__(self, 6)
```

Comment:
Because `f` and `s` are new classes, this class was previously directly instantiated with different parameters twice, and this class is used with yet a third parameter in tests. 

However, you could consider all of that historical baggage. As it's a simple refactoring, I'll go ahead and make some changes.